### PR TITLE
[TechDocs] Introduce the annotation `backstage.io/techdocs-ref: <relative-target>` as an alias for `backstage.io/techdocs-ref: dir:<relative-target>`

### PR DESCRIPTION
**Pull request by dhenneke**
Closes #6263
Relates to #4409

This PR is an outcome of the discussions around #6263. We want to make it easy to reference TechDocs documentation that is stored in the same folder as the entity definition. I first proposed `url:.`, which was a bit weird. Then @iamEAP proposed to stick with `dir:.`. Finally, @freben noted that it would also be cool to just write `backstage.io/techdocs-ref: .`. After some experimenting, I figured out that `backstage.io/techdocs-ref: ./` is just an alias for the longer `backstage.io/techdocs-ref: dir:./`. And that's what I implemented here 😀.

I cleaned up the original `dir` implementation to only support `url` and `file` targets. When we also drop support for the deprecated `github`, ... annotations we can also delete a lot of now unneeded helper functions. Note that `backstage.io/techdocs-ref: https://github.com...` is **not** supported and will throw an exception that targets the user to use `url:https://github.com...` instead.

I didn't change any documentation or examples yet. But I could add it to this PR, or we do a followup that focuses on this part.

---

Introduce the annotation `backstage.io/techdocs-ref: <relative-target>` as an alias for `backstage.io/techdocs-ref: dir:<relative-target>`. This annotation works with both the basic and the recommended flow, however, it will be most useful with the basic approach.

In addition, this change removes the support of the deprecated `github`, `gitlab`, and `azure/api` locations from the `dir` reference preparer.

#### Example Usage

The new annotation is convenient if the documentation is stored in the same location, i.e. the same git repository, as the `catalog-info.yaml`. While it is still supported to add full URLs such as `backstage.io/techdocs-ref: url:https://...` for custom setups, documentation is mostly stored in the same repository as the entity definition. By automatically resolving the target relative to the registration location of the entity, the configuration overhead for this default setup is minimized. Since it leverages the `@backstage/integrations` package for the URL resolution, this is compatible with every supported source.

Consider the following examples:

> Note that the short version `<target>` is only an alias for the still supported `dir:<target>`.
1. "I have a repository with a single `catalog-info.yaml` and a TechDocs page in the root folder!"

```
https://github.com/backstage/example/tree/main/
 |- catalog-info.yaml
 |  > apiVersion: backstage.io/v1alpha1
 |  > kind: Component
 |  > metadata:
 |  >   name: example
 |  >   annotations:
 |  >     backstage.io/techdocs-ref: . # -> same folder
 |  > spec: {}
 |- docs/
 |- mkdocs.yml
```

2. "I have a repository with a single `catalog-info.yaml` and my TechDocs page in located in a folder!"

```
https://bitbucket.org/my-owner/my-project/src/master/
 |- catalog-info.yaml
 |  > apiVersion: backstage.io/v1alpha1
 |  > kind: Component
 |  > metadata:
 |  >   name: example
 |  >   annotations:
 |  >     backstage.io/techdocs-ref: ./some-folder # -> subfolder
 |  > spec: {}
 |- some-folder/
   |- docs/
   |- mkdocs.yml
```

3. "I have a mono repository that hosts multiple components!"

```
https://dev.azure.com/organization/project/_git/repository
 |- my-1st-module/
   |- catalog-info.yaml
   |  > apiVersion: backstage.io/v1alpha1
   |  > kind: Component
   |  > metadata:
   |  >   name: my-1st-module
   |  >   annotations:
   |  >     backstage.io/techdocs-ref: . # -> same folder
   |  > spec: {}
   |- docs/
   |- mkdocs.yml
 |- my-2nd-module/
   |- catalog-info.yaml
   |  > apiVersion: backstage.io/v1alpha1
   |  > kind: Component
   |  > metadata:
   |  >   name: my-2nd-module
   |  >   annotations:
   |  >     backstage.io/techdocs-ref: . # -> same folder
   |  > spec: {}
   |- docs/
   |- mkdocs.yml
 |- catalog-info.yaml
 |  > apiVersion: backstage.io/v1alpha1
 |  > kind: Location
 |  > metadata:
 |  >   name: example
 |  > spec:
 |  >   targets:
 |  >     - ./*/catalog-info.yaml
```

All examples should work regardless of the SCM provider and even when the entity is registered from a `file` location in a development environment.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/techdocs-common/src/helpers.ts [hunk 1]
```diff
@@ -50,9 +60,113 @@ export const parseReferenceAnnotation = (
...
+        target,
+      };
+    }
+
+    default:
+      throw new InputError(`Unable to resolve location type ${location.type}`);
+  }
+};
+
 export const getLocationForEntity = (
```
Comment (dhenneke) [inline packages/techdocs-common/src/helpers.ts:161]: This function is also used to populate the location information to the `techdocs/metadata/entity/...` entpoint. This means the change is transparent to the UI as it only sees the resolved location.

**Review by OrkoHunter — APPROVED** [review]
Looks great! Two questions:

1. The changeset is great! But I feel these docs are worthy enough to be stored somewhere on backstage.io/docs. At the very least, we should update the pages mentioning this ref like
  - https://backstage.io/docs/features/techdocs/creating-and-publishing 
  - https://backstage.io/docs/features/techdocs/how-to-guides#how-to-use-url-reader-in-techdocs-prepare-step
  - https://backstage.io/docs/features/software-catalog/well-known-annotations#backstageiotechdocs-ref
2. I think this PR closes https://github.com/backstage/backstage/issues/4409 completely.
3. As a follow-up of this PR, can we update the default software templates to use `.` by default? https://github.com/backstage/software-templates

**Comment by dhenneke** [comment]
> Looks great!
> 
> 1. The changeset is great! But I feel these docs are worthy enough to be stored somewhere on backstage.io/docs. At the very least, we should update the pages mentioning this ref like
> 
> * https://backstage.io/docs/features/techdocs/creating-and-publishing
> * https://backstage.io/docs/features/techdocs/how-to-guides#how-to-use-url-reader-in-techdocs-prepare-step
> * https://backstage.io/docs/features/software-catalog/well-known-annotations#backstageiotechdocs-ref

Yes! Should I add it to this PR or do we want to do it in a follow-up?

> 2. I think this PR closes [TechDocs: Deprecate legacy `git clone` and dir preparer in favor of URL Reader #4409](https://github.com/backstage/backstage/issues/4409) completely.

There is still support for the old `github`, ... locations. I only removed them from the `dir:` location. But I can do a follow-up PR to close the issue, to also remove the [`commonGit.ts` file](https://github.com/backstage/backstage/blob/257b9d5693597e5fec58596147f139a44614b750/packages/techdocs-common/src/stages/prepare/commonGit.ts) and the corresponding registrations: https://github.com/backstage/backstage/blob/257b9d5693597e5fec58596147f139a44614b750/packages/techdocs-common/src/stages/prepare/preparers.ts#L50-L55

> 3. As a follow-up of this PR, can we update the default software templates to use `.` by default? https://github.com/backstage/software-templates

👍

**Comment by OrkoHunter** [comment]
@dhenneke I think we can create a docs PR separately and merge it along with the release this Thursday.

Thanks for your response on the other points :)

File: packages/techdocs-common/src/helpers.ts [hunk 2]
```diff
@@ -50,9 +60,113 @@ export const parseReferenceAnnotation = (
...
+  if (!annotation) {
+    return undefined;
+  }
+
+  if (typeof annotation !== 'string') {
+    return undefined;
+  }
+
+  // if the string doesn't contain `:`, interpret it as the target of a dir type
+  if (!annotation.includes(':')) {
```
Comment (iamEAP) [inline packages/techdocs-common/src/helpers.ts:83]: Hey @dhenneke, @ottosichert @camilaibs and I were taking a look at this and we're wondering if there's any utility in allowing a non-prefixed ref value beyond 4-5 saved characters?

File: packages/techdocs-common/src/helpers.test.ts [hunk 3]
```diff
@@ -109,10 +126,230 @@ describe('parseReferenceAnnotation', () => {
...
+describe('getDirLocation', () => {
+  it.each`
+    techdocsRef                    | responseTarget
+    ${undefined}                   | ${undefined}
+    ${16}                          | ${undefined}
+    ${'.'}                         | ${'.'}
+    ${'dir:.'}                     | ${'.'}
+    ${'./relative'}                | ${'./relative'}
+    ${'dir:./relative'}            | ${'./relative'}
+    ${'dir:https://github.com...'} | ${'https://github.com...'}
```
Comment (iamEAP) [inline packages/techdocs-common/src/helpers.test.ts:138]: Is this intended?  It doesn't seem like a valid use-case.

**Review by iamEAP — APPROVED** [review]
Thanks again @dhenneke!  This looks great overall; we just had a couple of questions where we wanted to understand the rationale.

File: packages/techdocs-common/src/helpers.test.ts (hunk 3 above)
Comment (dhenneke) [reply packages/techdocs-common/src/helpers.test.ts:138]: Yeah more or less. It _is_ possible to use it as long as your entity is registered by a `url`-type location, but I wouldn't recommended it. So we better remove it from the test. We could also add some special logic to recognize that it is a url and throw an error like in the `backstage.io/techdocs-ref: https://...` case, but I am not really keen of doing that since we can't reuse any logic as we did in the other case. After all, this value will probably lead to a rejection in a later stage. What do you think?

File: packages/techdocs-common/src/helpers.ts (hunk 2 above)
Comment (dhenneke) [reply packages/techdocs-common/src/helpers.ts:83]: Yes you can boil it down to just saving 4 characters. Considering that we can probably expect that `backstage.io/techdocs-ref: .` will be the default for the majority of use-cases, I find it _better_ from a user perspective. We should also consider that `<type>:<target>` is well-known across Backstage as it is also used in the catalog. **But**, a normal Backstage user will probably never need to understand that there are different types of locations where an entity might origin. They register their entities via the catalog-import or get it registered by the scaffolder, and don't need to write any location reference by hand. Even if they choose to use a [`Location` entity](https://backstage.io/docs/features/software-catalog/descriptor-format#kind-location), it has code to resolve [either relative or absolute locations](https://github.com/backstage/backstage/blob/20bb197b6f7fc4e7b1227b1c0ab7e47e90a4ca6e/plugins/catalog-backend/src/ingestion/processors/LocationEntityProcessor.ts#L23-L39) – it even doesn't support `<type>:<location>` at all. But they will need to do it for the `backstage.io/techdocs-ref` annotation and from this point-of-view, I would argue that it is more consistent if it also supports the short-form.

But there is also another perspective of this and this is "documentation" (i.e. of the feature). There are still some places that prefer `<type>:<target>` and have no automatic logic for a short form. However, this is primarily relevant for integrators. For them, it might be easier to understand `dir:.` (even though the catalog only supports `file`), and that this also implies that there might be other alternatives (-> `url`). [TechDocs - Creating and publishing your docs](https://backstage.io/docs/features/techdocs/creating-and-publishing) might be easier to read if it just lists (a) `dir` (recommended): relative location based on your entity definition file, and (b) `url` (advanced): absolute url to a location (e.g. a GitHub/GitLab/... repository). Compared to this, an additional paragraph "you can also skip the `dir:` if you like, but please never skip the `url:` since this will break your entity" might introduce confusion that might result in issues or support questions. A possible outcome of this argument could be to either skip the "short-form" or to make it compatible with URLs too.

Long story short: I would like to keep this feature but don't plan to add support for plain Urls. But I would also be fine with removing it and just keep the updated `dir:` behavior and the establishment of a new standard.

Comment (iamEAP) [reply packages/techdocs-common/src/helpers.ts:83]: > _But, a normal Backstage user will probably never need to understand that there are different types of locations where an entity might originate. They register their entities via the catalog-import or get it registered by the scaffolder, and don't need to write any location reference by hand._

Ah-hah.  Yeah, this resonates.  Backstage is probably overly guilty of this more broadly (leakage of knowledge/expectations between persona boundaries)...  So I like the idea of simplifying.

File: packages/techdocs-common/src/helpers.test.ts (hunk 3 above)
Comment (iamEAP) [reply packages/techdocs-common/src/helpers.test.ts:138]: Indeed, @dhenneke!  Yeah, perhaps it's fine to leave as-is here, but it would be great to handle this rejection in TechDocs backend.

Trying this value locally results in an anonymous `Error` in the UI and nothing in the build logs but `waiting for logs`.  Catching and forwarding the error when the `DocsBuilder` is instantiated in the `DocsSynchronizer` would close the loop, I think.

Comment (dhenneke) [reply packages/techdocs-common/src/helpers.test.ts:138]: > Catching and forwarding the error when the `DocsBuilder` is instantiated in the `DocsSynchronizer` would close the loop, I think.

Correct. I didn't thought about the possibility that the constructor can throw in #6341. When I move it into the `try {} catch {}` block I get the following result:

image

Not the fanciest error, but the user has at least a chance to deduce what was going on.

File: packages/techdocs-common/src/helpers.ts (hunk 2 above)
Comment (ottosichert) [reply packages/techdocs-common/src/helpers.ts:83]: Although allowing a short-hand notation seems like a minor aspect, we have taken the time to go through all implications of this, i.e.:

Pros
- Less to type, looks clean
- Backstage/TechDocs users don't need to be aware of `<type>:<target>` notation

Cons
- `url:https://example.com` is a valid (non-windows) folder name
- Two ways to specify directories: `./path` or `dir:./path` might cause confusion
- Implementation is more complex due to discovery of short-hand

Also, in the future the requirement of the ref might be completely dropped by using catalog processors, thus being more explicit at this point would be the best option to move forward in our point of view.

Comment (ottosichert) [reply packages/techdocs-common/src/helpers.ts:83]: What do you think @dhenneke?

Comment (dhenneke) [reply packages/techdocs-common/src/helpers.ts:83]: @ottosichert I think these are valid arguments and I agree that it will be best to remove the special handling of the short-hand notation. Should only be a few places and tests as the majority of code is concerning the actual `dir:` location. I'll update the PR.

**Comment by dhenneke** [comment]
I updated the PR to only make the `dir:<relative>` type annotation work properly. Thanks for your input @iamEAP @ottosichert @camilaibs @OrkoHunter!

Open Follow-ups:
* Update the documentation around it (see https://github.com/backstage/backstage/pull/6495#pullrequestreview-709526005).
* Use the `dir:.` annotation as a default in the templates.
* Remove the deprecations around #4409 and cleanup the package.
